File: sunflower-core/src/main/java/org/tdf/sunflower/controller/JsonRpc.java
package org.tdf.sunflower.controller;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.tdf.sunflower.types.Block;
import org.tdf.sunflower.types.LogInfo;
import org.tdf.sunflower.types.Transaction;
import org.tdf.sunflower.vm.CallData;
import org.tdf.sunflower.vm.CallType;

import java.util.Arrays;
import java.util.List;

import static org.tdf.sunflower.controller.TypeConverter.*;

public interface JsonRpc {

    String web3_clientVersion();

    String web3_sha3(String data) throws Exception;

    String net_version();

    String eth_chainId();

    String net_peerCount();

    boolean net_listening();

    String eth_protocolVersion();

    Object eth_syncing();

    String eth_coinbase();

    boolean eth_mining();

    String eth_hashrate();

    String eth_gasPrice();

    String[] eth_accounts();

    String eth_blockNumber();

    String eth_getBalance(String address, String block) throws Exception;

    String eth_getLastBalance(String address) throws Exception;

    String eth_getStorageAt(String address, String storageIdx, String blockId) throws Exception;

    String eth_getTransactionCount(String address, String blockId) throws Exception;

    String eth_getBlockTransactionCountByHash(String blockHash) throws Exception;

    String eth_getBlockTransactionCountByNumber(String bnOrId) throws Exception;

    String eth_getUncleCountByBlockHash(String blockHash) throws Exception;

    String eth_getUncleCountByBlockNumber(String bnOrId) throws Exception;

    String eth_getCode(String addr, String bnOrId) throws Exception;

    String eth_sign(String addr, String data) throws Exception;

    String eth_sendTransaction(CallArguments transactionArgs) throws Exception;

    String eth_sendRawTransaction(String rawData) throws Exception;

    String eth_call(CallArguments args, String bnOrId) throws Exception;

    String eth_estimateGas(CallArguments args) throws Exception;

    BlockResult eth_getBlockByHash(String blockHash, Boolean fullTransactionObjects) throws Exception;

    BlockResult eth_getBlockByNumber(String bnOrId, Boolean fullTransactionObjects) throws Exception;

    TransactionResultDTO eth_getTransactionByHash(String transactionHash) throws Exception;

    TransactionResultDTO eth_getTransactionByBlockHashAndIndex(String blockHash, String index) throws Exception;

    TransactionResultDTO eth_getTransactionByBlockNumberAndIndex(String bnOrId, String index) throws Exception;

    TransactionReceiptDTO eth_getTransactionReceipt(String transactionHash) throws Exception;

    TransactionReceiptDTOExt ethj_getTransactionReceipt(String transactionHash) throws Exception;

    BlockResult eth_getUncleByBlockHashAndIndex(String blockHash, String uncleIdx) throws Exception;

    BlockResult eth_getUncleByBlockNumberAndIndex(String blockId, String uncleIdx) throws Exception;


    String eth_newFilter(FilterRequest fr) throws Exception;

    String eth_newBlockFilter();

    String eth_newPendingTransactionFilter();

    boolean eth_uninstallFilter(String id);

    Object[] eth_getFilterChanges(String id);

    Object[] eth_getFilterLogs(String id);

    Object[] eth_getLogs(FilterRequest fr) throws Exception;

//    String eth_resend();
//    String eth_pendingTransactions();

    List<Object> eth_getWork();

//    String eth_newFilter(String fromBlock, String toBlock, String address, String[] topics) throws Exception;

    boolean eth_submitWork(String nonce, String header, String digest) throws Exception;

    boolean eth_submitHashrate(String hashrate, String id);


    String personal_signAndSendTransaction(CallArguments tx, String password);
//    boolean miner_startAutoDAG();
//    boolean miner_stopAutoDAG();
//    boolean miner_makeDAG();
//    String miner_hashrate();

//    String debug_printBlock();
//    String debug_getBlockRlp();
//    String debug_setHead();
//    String debug_processBlock();
//    String debug_seedHash();
//    String debug_dumpBlock();
//    String debug_metrics();

    @Value
    @AllArgsConstructor
    @ToString
    class SyncingResult {
        private final String startingBlock;
        private final String currentBlock;
        private final String highestBlock;
    }

    @NoArgsConstructor
    @AllArgsConstructor
    class CallArguments {
        public String from;
        public String to;
        public String gas;
        public String gasPrice;
        public String value;
        public String data; // compiledCode
        public String nonce;

        @JsonIgnore
        @SneakyThrows
        public CallData toCallData() {
            CallData data = CallData.empty();
            if (from != null && !from.isEmpty()) {
                data.setCaller(jsonHexToHexBytes(from));
                data.setOrigin(jsonHexToHexBytes(from));
            }
            if (to != null && !to.isEmpty()) {
                data.setTxTo(jsonHexToHexBytes(to));
                data.setTo(jsonHexToHexBytes(to));
                data.setCallType(CallType.CALL);
            } else {
                data.setCallType(CallType.CREATE);
            }
            if (gas != null && !gas.isEmpty()) {
                data.setGasLimit(
                    jsonHexToU256(gas)
                );
            }
            if (gasPrice != null && !gasPrice.isEmpty()) {
                data.setGasPrice(
                    jsonHexToU256(gasPrice)
                );
            }
            if (value != null && !value.isEmpty()) {
                data.setValue(
                    jsonHexToU256(value)
                );
                data.setTxValue(
                    jsonHexToU256(value)
                );
            }
            if (this.data != null && !this.data.isEmpty()) {
                data.setData(
                    jsonHexToHexBytes(this.data)
                );
            }
            if (nonce != null && !nonce.isEmpty()) {
                data.setTxNonce(jsonHexToLong(nonce));
            }
            return data;
        }

        @Override
        public String toString() {
            return "CallArguments{" +
                "from='" + from + '\'' +
                ", to='" + to + '\'' +
                ", gas='" + gas + '\'' +
                ", gasPrice='" + gasPrice + '\'' +
                ", value='" + value + '\'' +
                ", data='" + data + '\'' +
                ", nonce='" + nonce + '\'' +
                '}';
        }
    }

    class BlockResult {
        public String number; // QUANTITY - the block number. null when its pending block.
        public String hash; // DATA, 32 Bytes - hash of the block. null when its pending block.
        public String parentHash; // DATA, 32 Bytes - hash of the parent block.
        public String nonce; // DATA, 8 Bytes - hash of the generated proof-of-work. null when its pending block.
        public String sha3Uncles; // DATA, 32 Bytes - SHA3 of the uncles data in the block.
        public String logsBloom; // DATA, 256 Bytes - the bloom filter for the logs of the block. null when its pending block.
        public String transactionsRoot; // DATA, 32 Bytes - the root of the transaction trie of the block.
        public String stateRoot; // DATA, 32 Bytes - the root of the final state trie of the block.
        public String receiptsRoot; // DATA, 32 Bytes - the root of the receipts trie of the block.
        public String miner; // DATA, 20 Bytes - the address of the beneficiary to whom the mining rewards were given.
        public String difficulty; // QUANTITY - integer of the difficulty for this block.
        public String totalDifficulty; // QUANTITY - integer of the total difficulty of the chain until this block.
        public String extraData; // DATA - the "extra data" field of this block
        public String size;//QUANTITY - integer the size of this block in bytes.
        public String gasLimit;//: QUANTITY - the maximum gas allowed in this block.
        public String gasUsed; // QUANTITY - the total used gas by all transactions in this block.
        public String timestamp; //: QUANTITY - the unix timestamp for when the block was collated.
        public Object[] transactions; //: Array - Array of transaction objects, or 32 Bytes transaction hashes depending on the last given parameter.
        public String[] uncles; //: Array - Array of uncle hashes.

        @Override
        public String toString() {
            return "BlockResult{" +
                "number='" + number + '\'' +
                ", hash='" + hash + '\'' +
                ", parentHash='" + parentHash + '\'' +
                ", nonce='" + nonce + '\'' +
                ", sha3Uncles='" + sha3Uncles + '\'' +
                ", logsBloom='" + logsBloom + '\'' +
                ", transactionsRoot='" + transactionsRoot + '\'' +
                ", stateRoot='" + stateRoot + '\'' +
                ", receiptsRoot='" + receiptsRoot + '\'' +
                ", miner='" + miner + '\'' +
                ", difficulty='" + difficulty + '\'' +
                ", totalDifficulty='" + totalDifficulty + '\'' +
                ", extraData='" + extraData + '\'' +
                ", size='" + size + '\'' +
                ", gas='" + gasLimit + '\'' +
                ", gasUsed='" + gasUsed + '\'' +
                ", timestamp='" + timestamp + '\'' +
                ", transactions=" + Arrays.toString(transactions) +
                ", uncles=" + Arrays.toString(uncles) +
                '}';
        }
    }

    class FilterRequest {
        public String fromBlock;
        public String toBlock;
        public Object address;
        public Object[] topics;
        public String blockHash;  // EIP-234: makes fromBlock = toBlock = blockHash

        @Override
        public String toString() {
            return "FilterRequest{" +
                "fromBlock='" + fromBlock + '\'' +
                ", toBlock='" + toBlock + '\'' +
                ", address=" + address +
                ", topics=" + Arrays.toString(topics) +
                ", blockHash='" + blockHash + '\'' +
                '}';
        }
    }

    class LogFilterElement {
        public String logIndex;
        public String transactionIndex;
        public String transactionHash;
        public String blockHash;
        public String blockNumber;
        public String address;
        public String data;
        public String[] topics;

        public LogFilterElement(LogInfo logInfo, Block b, Integer txIndex, Transaction tx, int logIdx) {
            logIndex = toJsonHex(logIdx);
            blockNumber = b == null ? null : toJsonHex(b.getHeight());
            blockHash = b == null ? null : toJsonHex(b.getHash().getBytes());
            transactionIndex = b == null ? null : toJsonHex(txIndex);
            transactionHash = toJsonHex(tx.getHash());
            address = toJsonHex(tx.getReceiveAddress());
            data = toJsonHex(logInfo.getData());
            topics = new String[logInfo.getTopics().size()];
            for (int i = 0; i < topics.length; i++) {
                topics[i] = toJsonHex(logInfo.getTopics().get(i).getData());
            }
        }

        @Override
        public String toString() {
            return "LogFilterElement{" +
                "logIndex='" + logIndex + '\'' +
                ", blockNumber='" + blockNumber + '\'' +
                ", blockHash='" + blockHash + '\'' +
                ", transactionHash='" + transactionHash + '\'' +
                ", transactionIndex='" + transactionIndex + '\'' +
                ", address='" + address + '\'' +
                ", data='" + data + '\'' +
                ", topics=" + Arrays.toString(topics) +
                '}';
        }
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/controller/JsonRpc.kt
package org.tdf.sunflower.controller

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.databind.JsonNode
import org.tdf.common.util.HexBytes
import org.tdf.sunflower.controller.JsonRpcUtil.toCallData
import org.tdf.sunflower.state.AddrUtil
import org.tdf.sunflower.types.Block
import org.tdf.sunflower.types.LogInfo
import org.tdf.sunflower.types.Transaction
import org.tdf.sunflower.vm.CallContext
import org.tdf.sunflower.vm.CallData

interface JsonRpc {
    fun web3_clientVersion(): String

    fun web3_sha3(data: String): String
    fun net_version(): String
    fun eth_chainId(): String
    fun net_peerCount(): String
    fun net_listening(): Boolean
    fun eth_protocolVersion(): String
    fun eth_syncing(): Boolean
    fun eth_coinbase(): String
    fun eth_mining(): Boolean
    fun eth_hashrate(): String
    fun eth_gasPrice(): String
    fun eth_accounts(): Array<String>
    fun eth_blockNumber(): String

    fun eth_getBalance(address: String, block: String): String

    fun eth_getStorageAt(address: String, storageIdx: String, blockId: String): String

    fun eth_getTransactionCount(address: String, blockId: String): String

    fun eth_getBlockTransactionCountByHash(blockHash: String): String?

    fun eth_getBlockTransactionCountByNumber(bnOrId: String): String?

    fun eth_getUncleCountByBlockHash(blockHash: String): String

    fun eth_getUncleCountByBlockNumber(bnOrId: String): String

    fun eth_getCode(addr: String, bnOrId: String): String

    fun eth_sign(addr: String, data: String): String

    fun eth_sendTransaction(transactionArgs: CallArguments): String

    fun eth_sendRawTransaction(rawData: String): String

    fun eth_call(args: CallArguments, bnOrId: String): String

    fun eth_estimateGas(args: CallArguments): String

    fun eth_getBlockByHash(blockHash: String, fullTx: Boolean?): BlockResult?

    fun eth_getBlockByNumber(bnOrId: String, fullTx: Boolean?): BlockResult?


    fun eth_getTransactionByHash(transactionHash: String): TransactionResultDTO?


    fun eth_getTransactionByBlockHashAndIndex(blockHash: String, index: String): TransactionResultDTO?


    fun eth_getTransactionByBlockNumberAndIndex(bnOrId: String, index: String): TransactionResultDTO?


    fun eth_getTransactionReceipt(transactionHash: String): TransactionReceiptDTO?


    fun eth_getUncleByBlockHashAndIndex(blockHash: String?, uncleIdx: String?): BlockResult?


    fun eth_getUncleByBlockNumberAndIndex(blockId: String?, uncleIdx: String?): BlockResult?


    fun eth_newFilter(fr: FilterRequest?): String?
    fun eth_newBlockFilter(): String?
    fun eth_newPendingTransactionFilter(): String?
    fun eth_uninstallFilter(id: String?): Boolean
    fun eth_getFilterChanges(id: String?): Array<Any?>?
    fun eth_getFilterLogs(id: String?): Array<Any?>?


    fun eth_getLogs(fr: FilterRequest): List<Any>

    //    String eth_resend();
    //    String eth_pendingTransactions();
    fun eth_getWork(): List<Any?>?

    //    String eth_newFilter(String fromBlock, String toBlock, String address, String[] topics) throws Exception;

    fun eth_submitWork(nonce: String?, header: String?, digest: String?): Boolean
    fun eth_submitHashrate(hashrate: String?, id: String?): Boolean
    fun personal_signAndSendTransaction(tx: CallArguments?, password: String?): String?

    data class CallArguments(
        val from: String? = null,
        val to: String? = null,
        val gas: String? = null,
        val gasPrice: String? = null,
        val value: String? = null,
        val data // compiledCode
        : String? = null,
        val nonce: String? = null,
    ) {
        @JsonIgnore
        fun toCallContext(
            nonce: Long?,
            chainId: Int,
            timestamp: Long = System.currentTimeMillis() / 1000,
            coinbase: HexBytes = AddrUtil.empty(),
            blockHashMap: Map<Long, ByteArray> = emptyMap()
        ): CallContext {
            return JsonRpcUtil.toCallContext(this, nonce, chainId, timestamp, coinbase, blockHashMap)
        }

        @JsonIgnore
        fun toCallData(): CallData {
            return toCallData(this)
        }
    }

    data class BlockResult(
        val number // QUANTITY - the block number. null when its pending block.
        : String? = null,

        val hash // DATA, 32 Bytes - hash of the block. null when its pending block.
        : String? = null,

        val parentHash // DATA, 32 Bytes - hash of the parent block.
        : String? = null,

        val nonce // DATA, 8 Bytes - hash of the generated proof-of-work. null when its pending block.
        : String? = null,

        val sha3Uncles // DATA, 32 Bytes - SHA3 of the uncles data in the block.
        : String? = null,

        val logsBloom // DATA, 256 Bytes - the bloom filter for the logs of the block. null when its pending block.
        : String? = null,

        val transactionsRoot // DATA, 32 Bytes - the root of the transaction trie of the block.
        : String? = null,

        val stateRoot // DATA, 32 Bytes - the root of the final state trie of the block.
        : String? = null,
        val receiptsRoot // DATA, 32 Bytes - the root of the receipts trie of the block.
        : String? = null,
        val miner // DATA, 20 Bytes - the address of the beneficiary to whom the mining rewards were given.
        : String? = null,
        val difficulty // QUANTITY - integer of the difficulty for this block.
        : String? = null,
        val totalDifficulty // QUANTITY - integer of the total difficulty of the chain until this block.
        : String? = null,
        val extraData // DATA - the "extra data" field of this block
        : String? = null,
        val size //QUANTITY - integer the size of this block in bytes.
        : String? = null,
        val gasLimit //: QUANTITY - the maximum gas allowed in this block.
        : String? = null,
        val gasUsed // QUANTITY - the total used gas by all transactions in this block.
        : String? = null,
        val timestamp //: QUANTITY - the unix timestamp for when the block was collated.
        : String? = null,
        val transactions //: Array - Array of transaction objects, or 32 Bytes transaction hashes depending on the last given parameter.
        : List<Any>? = null,
        val uncles //: Array - Array of uncle hashes.
        : List<String>? = null,

        var mixHash: String? = null,
    )


    data class FilterRequest(
        val fromBlock: String? = null,
        val toBlock: String? = null,
        val address: JsonNode? = null,
        val topics: List<JsonNode?>? = null,
        val blockHash // EIP-234: makes fromBlock = toBlock = blockHash
        : String? = null,
    )

    data class LogFilterElement(
        val logIndex: String? = null,
        val transactionIndex: String? = null,
        val transactionHash: String? = null,
        val blockHash: String? = null,
        val blockNumber: String? = null,
        val address: String? = null,
        val data: String? = null,
        val topics: List<String>? = null
    ) {
        companion object {
            fun create(info: LogInfo, b: Block?, txIndex: Int?, tx: Transaction, logIdx: Int): LogFilterElement {
                return LogFilterElement(
                    logIdx.jsonHex,
                    txIndex?.jsonHex,
                    tx.hash.jsonHex,
                    b?.hash?.jsonHex,
                    b?.height?.jsonHex,
                    tx.to.takeIf { !it.isEmpty() }?.jsonHex ?: tx.contractAddress?.jsonHex,
                    info.data.jsonHex,
                    info.topics.map { it.jsonHex }
                )
            }
        }
    }
}